src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs

```diff
@@ -92,21 +94,16 @@ toVar srcs c = fromBinary $ castBits $ binaryExpansion ex
         x  = g ^ runBinScale (v `evalPolynomial` c)
         ex = foldr (\p y -> x ^ p + y) x srcs
 
-con2var :: Arithmetic a => Constraint a -> Integer
-con2var = toVar []
-
-newVariable :: State (ArithmeticCircuit a) Integer
-newVariable = do
-    r <- get
-    let (x, g) = randomR (0, order @VarField - 1) (acRNG r)
-    put r { acRNG = g }
-    return x
-
 newVariableWithSource :: Arithmetic a => [Integer] -> (Integer -> Constraint a) -> State (ArithmeticCircuit a) Integer
-newVariableWithSource srcs con = toVar srcs . con <$> newVariable
+newVariableWithSource srcs con = toVar srcs . con . fst <$> do
+    zoom #acRNG $ get >>= traverse put . randomR (0, order @VarField - 1)
 
-addVariable :: Integer -> State (ArithmeticCircuit a) ()
-addVariable x = modify (\r -> r { acOutput = x, acVarOrder = insert (length (acVarOrder r), x) x (acVarOrder r)})
+addVariable :: Integer -> State (ArithmeticCircuit a) Integer
+addVariable x = do
+    zoom #acOutput $ put x
+    zoom #acVarOrder . modify
+        $ insert <$> (,x) . length <*> const x <*> id
```

Comment:
I would rather introduce a new variable for `acVarOrder` argument, it would be a little easier to read